### ISSUE TITLE
infer结果输出值固定不变，但是训练中的test pass显示的错误率却是变化的

### ISSUE THREAD

**Issue by xiangjinxin1019** [PaddlePaddle/Paddle#5574]
参考/book 中的手写数字识别demo，用的是自己的图片数据集，只分2类（后缀名含correct的图片的label=0；后缀名含incorrect的图片的label=1），也自己定义了一个reader_creator()，代码如下。训练过程中，test pass 显示的错误率一直在50%附近波动，最后也手动预测了自己测试集中的800张图片，预测的结果全部是固定值，label=1。是不是训练根本没有起作用？还是预测那块有问题？代码如下，辛苦哪位大神帮忙看下

**Comment by xiangjinxin1019** [comment]
这是自定义的reader_creator，训练集文件夹中存放有n张 100*100的图片，遍历文件夹，Image.open每张图片，根据后缀名('correct' or 'incorrect')取label, 然后yield image和label，最后return reader

def image_reader_creator(folder_path):
    def reader():
        for img in os.listdir(folder_path):
            if img.startswith('.') and os.path.isfile(os.path.join(folder_path, img)):
                os.listdir(folder_path).remove(img)
            img_path = folder_path + '/' + img
            label = None
            label_str = img.split('.')[0]
            if label_str == 'correct':
                label = 0
            elif label_str == 'incorrect':
                label = 1
            else:
                print 'label_str error:', label_str
                raise TypeError
            im = Image.open(img_path).convert('L')
            im = im.resize((100, 100), Image.ANTIALIAS)
            im = numpy.array(im).astype(numpy.float32).flatten()
            image = im / 255.0
            yield image, label  # a single entry of data is created each time

    return reader

**Comment by xiangjinxin1019** [comment]
import os
from PIL import Image
import numpy as np
import img_reader
import paddle.v2 as paddle

with_gpu = os.getenv('WITH_GPU', '0') != '0'


def convolutional_neural_network(img):
    # first conv layer
    conv_pool_1 = paddle.networks.simple_img_conv_pool(
        input=img,
        filter_size=5,
        num_filters=20,
        num_channel=1,
        pool_size=2,
        pool_stride=2,
        act=paddle.activation.Relu())
    # second conv layer
    conv_pool_2 = paddle.networks.simple_img_conv_pool(
        input=conv_pool_1,
        filter_size=5,
        num_filters=50,
        num_channel=20,
        pool_size=2,
        pool_stride=2,
        act=paddle.activation.Relu())
    # fully-connected layer
    predict = paddle.layer.fc(
        input=conv_pool_2, size=2, act=paddle.activation.Softmax())
    return predict


def main():
    paddle.init(use_gpu=False, trainer_count=1)

    # define network topology
    images = paddle.layer.data(
        name='image', type=paddle.data_type.dense_vector(100 * 100))
    label = paddle.layer.data(
        name='label', type=paddle.data_type.integer_value(2))

    predict = convolutional_neural_network(images)

    cost = paddle.layer.classification_cost(input=predict, label=label)

    parameters = paddle.parameters.create(cost)

    optimizer = paddle.optimizer.Momentum(
        learning_rate=0.1 / 128.0,
        momentum=0.9,
        regularization=paddle.optimizer.L2Regularization(rate=0.0005 * 128))

    trainer = paddle.trainer.SGD(
        cost=cost, parameters=parameters, update_equation=optimizer)

    lists = []

    def event_handler(event):
        if isinstance(event, paddle.event.EndIteration):
            if event.batch_id % 100 == 0:
                print "Pass %d, Batch %d, Cost %f, %s" % (
                    event.pass_id, event.batch_id, event.cost, event.metrics)
        if isinstance(event, paddle.event.EndPass):
            # save parameters
            with open('./output/params_pass_%d.tar' % event.pass_id, 'w') as f:
                # trainer.save_parameter_to_tar(f)
                parameters.to_tar(f)
	    cur_path = os.getcwd()
	    test_folder_path = os.path.join(cur_path, 'dataset/small/test') 
	    cus_test_reader = img_reader.image_reader_creator(test_folder_path)
            result = trainer.test(
			reader=paddle.batch(
				paddle.reader.shuffle(cus_test_reader, buf_size=1000), 
				128),
			feeding={"image":0, "label":1},
			)
            print "Test with Pass %d, Cost %f, %s\n" % (event.pass_id, result.cost, result.metrics)
            lists.append((event.pass_id, result.cost,
                          result.metrics['classification_error_evaluator']))
    
    cur_path = os.getcwd()
    train_folder_path = os.path.join(cur_path, 'dataset/small/train')
    cus_train_reader = img_reader.image_reader_creator(train_folder_path)
    trainer.train(
	reader=paddle.batch(
		paddle.reader.shuffle(cus_train_reader, buf_size=8000), 
		batch_size=128), 
	feeding={"image":0, "label":1},
	event_handler=event_handler,
	num_passes=100
	)

    # find the best pass
    best = sorted(lists, key=lambda list: float(list[1]))[0]
    print 'Best pass is %s, testing Avgcost is %s' % (best[0], best[1])
    print 'The classification accuracy is %.2f%%' % (100 - float(best[2]) * 100)

    def load_image(file):
        im = Image.open(file).convert('L')
        im = im.resize((100, 100), Image.ANTIALIAS)
        im = np.array(im).astype(np.float32).flatten()
        im = im / 255.0
        return im

    test_data = []
    cur_dir = os.path.dirname(os.path.realpath(__file__))
    test_data.append((load_image(cur_dir + '/dataset/small/train/incorrect.980.jpg'), ))

    probs = paddle.infer(
        output_layer=predict, parameters=parameters, input=test_data)
    lab = np.argsort(-probs)  # probs and lab are the results of one batch data
    print "Label of image:incorrect.980.jpg is: %d" % lab[0][0]


if __name__ == '__main__':
    main()

**Comment by xiangjinxin1019** [comment]
这是train.py

import os
from PIL import Image
import numpy as np
import img_reader
import paddle.v2 as paddle

with_gpu = os.getenv('WITH_GPU', '0') != '0'


def convolutional_neural_network(img):
    # first conv layer
    conv_pool_1 = paddle.networks.simple_img_conv_pool(
        input=img,
        filter_size=5,
        num_filters=20,
        num_channel=1,
        pool_size=2,
        pool_stride=2,
        act=paddle.activation.Relu())
    # second conv layer
    conv_pool_2 = paddle.networks.simple_img_conv_pool(
        input=conv_pool_1,
        filter_size=5,
        num_filters=50,
        num_channel=20,
        pool_size=2,
        pool_stride=2,
        act=paddle.activation.Relu())
    # fully-connected layer
    predict = paddle.layer.fc(
        input=conv_pool_2, size=2, act=paddle.activation.Softmax())
    return predict


def main():
    paddle.init(use_gpu=False, trainer_count=1)

    # define network topology
    images = paddle.layer.data(
        name='image', type=paddle.data_type.dense_vector(100 * 100))
    label = paddle.layer.data(
        name='label', type=paddle.data_type.integer_value(2))

    predict = convolutional_neural_network(images)

    cost = paddle.layer.classification_cost(input=predict, label=label)

    parameters = paddle.parameters.create(cost)

    optimizer = paddle.optimizer.Momentum(
        learning_rate=0.1 / 128.0,
        momentum=0.9,
        regularization=paddle.optimizer.L2Regularization(rate=0.0005 * 128))

    trainer = paddle.trainer.SGD(
        cost=cost, parameters=parameters, update_equation=optimizer)

    lists = []

    def event_handler(event):
        if isinstance(event, paddle.event.EndIteration):
            if event.batch_id % 100 == 0:
                print "Pass %d, Batch %d, Cost %f, %s" % (
                    event.pass_id, event.batch_id, event.cost, event.metrics)
        if isinstance(event, paddle.event.EndPass):
            # save parameters
            with open('./output/params_pass_%d.tar' % event.pass_id, 'w') as f:
                # trainer.save_parameter_to_tar(f)
                parameters.to_tar(f)
	    cur_path = os.getcwd()
	    test_folder_path = os.path.join(cur_path, 'dataset/small/test') 
	    cus_test_reader = img_reader.image_reader_creator(test_folder_path)
            result = trainer.test(
			reader=paddle.batch(
				paddle.reader.shuffle(cus_test_reader, buf_size=1000), 
				128),
			feeding={"image":0, "label":1},
			)
            print "Test with Pass %d, Cost %f, %s\n" % (event.pass_id, result.cost, result.metrics)
            lists.append((event.pass_id, result.cost,
                          result.metrics['classification_error_evaluator']))
    
    cur_path = os.getcwd()
    train_folder_path = os.path.join(cur_path, 'dataset/small/train')
    cus_train_reader = img_reader.image_reader_creator(train_folder_path)
    trainer.train(
	reader=paddle.batch(
		paddle.reader.shuffle(cus_train_reader, buf_size=8000), 
		batch_size=128), 
	feeding={"image":0, "label":1},
	event_handler=event_handler,
	num_passes=100
	)

    # find the best pass
    best = sorted(lists, key=lambda list: float(list[1]))[0]
    print 'Best pass is %s, testing Avgcost is %s' % (best[0], best[1])
    print 'The classification accuracy is %.2f%%' % (100 - float(best[2]) * 100)

    def load_image(file):
        im = Image.open(file).convert('L')
        im = im.resize((100, 100), Image.ANTIALIAS)
        im = np.array(im).astype(np.float32).flatten()
        im = im / 255.0
        return im

    test_data = []
    cur_dir = os.path.dirname(os.path.realpath(__file__))
    test_data.append((load_image(cur_dir + '/dataset/small/train/incorrect.980.jpg'), ))

    probs = paddle.infer(
        output_layer=predict, parameters=parameters, input=test_data)
    lab = np.argsort(-probs)  # probs and lab are the results of one batch data
    print "Label of image:incorrect.980.jpg is: %d" % lab[0][0]


if __name__ == '__main__':
    main()

**Comment by NHZlX** [comment]
`test pass 显示的错误率一直在50%附近波动`,   2个类别的话， 如果为random的输出，也是0.5的几率，所以网络没有收敛，请从以下两方面入手： 1. 确保数据没有问题，2.网络的设计（你的网络一共三层， 2 conv + 1fc， 认为这个的原因比较大）

**Comment by xiangjinxin1019** [comment]
请问针对我这个使用场景，网络一般怎么设计比较好？还有optimizer中的这三个参数，具体是对什么起作用的，可以在哪查到这些参数的说明

**Comment by xiangjinxin1019** [comment]
predict.py 中 infer的代码如下：
但infer输出的结果固定是1，不知道为啥


`test_data = []
    cur_dir = os.path.dirname(os.path.realpath(__file__))
    test_dir = os.path.join(cur_dir, 'dataset/small/test')

    with open('./output/params_pass_11.tar', 'r') as f:
        parameters = paddle.parameters.Parameters.from_tar(f)

    for test_img in os.listdir(test_dir):
        test_data.append((load_image(test_dir + '/' + test_img), ))

        probs = paddle.infer(
            output_layer=predict,
            parameters=parameters,
            input=test_data,
        )
        lab = np.argsort(-probs)  # probs and lab are the results of one batch data
        print 'Label of image:', test_img, "is: %d" % lab[0][0]`
